iwyu.cc

```diff
@@ -3852,12 +3861,9 @@ class IwyuAstConsumer
     if (!CanForwardDeclareType(current_ast_node())) {
       const map<const Type*, const Type*> resugar_map
           = GetTplTypeResugarMapForClass(type);
-      // ScanInstantiatedType requires that type not already be on the
-      // ast-stack that it sees, but in our case it will be.  So we
-      // pass in the parent-node.
-      const ASTNode* ast_parent = current_ast_node()->parent();
-      instantiated_template_visitor_.ScanInstantiatedType(
-          type, ast_parent, resugar_map);
+
+      instantiated_template_visitor_.ScanInstantiatedType(current_ast_node(),
```

Comment:
It is OK because I changed *ScanInstantiatedType* to expect it. Before the change, that function would call *TraverseType* which is hooked by IWYU to automatically push an *ASTNode* to the top of the stack (the *parent node*) My change is to call directly *TraverseTemplateSpecializationType* on *the current* ASTNode instead, so that it is in the caller context.

For example, this is how the AST stack looked before my change upon construction of a `DefaultArgument<int>`:
```
 - [CXXConstructExpr] CXXConstructExpr 0x2f34af8 <tests/cxx/template_default_args_roundtrip-template.h:14:10> 'struct DefaultArgument<int>':'struct DefaultArgument<int>' 'void (voi
 in tests/cxx/template_default_args_roundtrip-template.h:14:10
 - [CXXConstructorDecl] inline constexpr Base() noexcept in tests/cxx/template_default_args_roundtrip-template.h:14:10
 - [CXXRecordDecl] struct Base {} in tests/cxx/template_default_args_roundtrip-template.h:14:10
 - [TemplateSpecializationTypeLoc] Type<struct DefaultArgument<int> > in tests/cxx/template_default_args_roundtrip-template.h:25:12
 - [NestedNameSpecifier] Type<struct DefaultArgument<int> >:: in tests/cxx/template_default_args_roundtrip-template.h:25:12
 - [ElaboratedTypeLoc] typename Type<struct DefaultArgument<int> >::type in tests/cxx/template_default_args_roundtrip-template.h:25:3
 - [FieldDecl] typename Type<DefaultArgument<int> >::type t in tests/cxx/template_default_args_roundtrip-template.h:25:26
 - [TemplateSpecializationType] Template<int> in tests/cxx/template_default_args_roundtrip.cc:19:15
 C [VarDecl] Template<int> templ in tests/cxx/template_default_args_roundtrip.cc:19:15
 C [TranslationUnitDecl]  in Invalid location
```

The nodes marked with a C are those that are added before entering the *Scan..* functions (*C*aller context)  and the ones above are created by the template visitor. As you can see the template instantiation type is regarded as part of template code, but it should be part of the caller because its location is outside of the templates being instantiated.


After the fix, the stack looks like this:
```
 - [CXXConstructExpr] CXXConstructExpr 0x3a41af8 <tests/cxx/template_default_args_roundtrip-template.h:14:10> 'struct DefaultArgument<int>':'struct DefaultArgument<int>' 'void (voi in tests/cxx/template_default_args_roundtrip-template.h:14:10
 - [CXXConstructorDecl] inline constexpr Base() noexcept in tests/cxx/template_default_args_roundtrip-template.h:14:10
 - [CXXRecordDecl] struct Base {} in tests/cxx/template_default_args_roundtrip-template.h:14:10
 - [TemplateSpecializationTypeLoc] Type<struct DefaultArgument<int> > in tests/cxx/template_default_args_roundtrip-template.h:25:12
 - [NestedNameSpecifier] Type<struct DefaultArgument<int> >:: in tests/cxx/template_default_args_roundtrip-template.h:25:12
 - [ElaboratedTypeLoc] typename Type<struct DefaultArgument<int> >::type in tests/cxx/template_default_args_roundtrip-template.h:25:3
 - [FieldDecl] typename Type<DefaultArgument<int> >::type t in tests/cxx/template_default_args_roundtrip-template.h:25:26
 C [TemplateSpecializationTypeLoc] Template<int> in tests/cxx/template_default_args_roundtrip.cc:19:1
 C [VarDecl] Template<int> templ in tests/cxx/template_default_args_roundtrip.cc:19:15
 C [TranslationUnitDecl]  in Invalid location
```

Now, the specialization type (a typeloc instead) is correctly part of the caller stack, and only the actual template code is above.

I added lots of logging in a temporary branch to aid my way through (impossible otherwise), I can share it if you want.

Reply:
This is a nice way of thinking about it (which node belongs in which context). I think it makes sense.